web/js/codeworld.js

```diff
@@ -221,6 +221,59 @@ function initCodeworld() {
         }
     });
     window.codeworldEditor.refresh();
+    window.codeworldEditor.on("cursorActivity", function() {
+        const prevDiv = document.getElementById("function-details");
+        if (prevDiv) prevDiv.remove();
+
+        const enclosingFunction = getEnclosingFunction(window.codeworldEditor);
+
+        if (!enclosingFunction || !enclosingFunction.string ||  !window.codeWorldSymbols[enclosingFunction.string]) {
+            return;
+        }
+        const cursor = window.codeworldEditor.getCursor();
+        const currentToken = window.codeworldEditor.getTokenAt(cursor);
+        const lineTokens = window.codeworldEditor.getLineTokens(cursor.line);
+        const argIndex =
+            lineTokens.filter(
+                elem =>
+                    getEnclosingFunction(
+                        window.codeworldEditor,
+                        {
+                            line: cursor.line, ch: elem.end
+                        })
+                        .string === enclosingFunction.string
+                    && elem.type && elem.type !== "keyword"
+                    && elem.type!=="bracket-0"
+                    && elem.type !== "built-in"
+            ).findIndex(elem => (elem.start === currentToken.start || elem.end === currentToken.start || elem.start === currentToken.end || elem.end === currentToken.end));
+        const topDiv = document.createElement('div');
+
+        topDiv.title = enclosingFunction.string;
+        topDiv.id = "function-details";
+        const keywordData = window.codeWorldSymbols[enclosingFunction.string];
+
+        const docDiv = document.createElement('div');
+        docDiv.classList.add("function-tooltip-styling");
+
+        const annotation = document.createElement('div');
+        renderDeclaration(annotation, enclosingFunction.string, keywordData, 9999, argIndex);
+        annotation.className = 'hover-decl';
+        docDiv.appendChild(annotation);
+
+        topDiv.appendChild(docDiv);
+        window.codeworldEditor.addWidget(cursor, topDiv, true, "above", "left");
+    });
+
+    window.getEnclosingFunction = (cm, cur = cm.getCursor()) => {
```

Comment:
I'm pretty sure this is insufficient.  For example, consider a cursor right after `sin(x) + `.  Won't this tell you that `sin` is the enclosing function?  You actually need to scan backwards and match brackets until the last unmatched open bracket.

Another option (and my first instinct is that it would be better) would be to compute this during lexing in codeworld-mode.js, and store it in the token state.  Then you don't have to do so much rescanning of the text when the cursor moves.  CodeMirror would take care of all the logic needed to keep that state up-to-date as the document changes.  And your interpretation of the source code would be more consistent with syntax highlighting -- for example, when something like layout rules force the editor to conclude that there's an unmatched bracket, this code would be consistent with that assumption.  You'd also get the feature to work for line-wrapped functions, for free.

Reply:
looking over this method, if it doesn't work out I guess I will go with the method I mentioned above.

Reply:
@cdsmith I have implemented your approach and it works like a charm. Please have a look :)